source/common/unified_malloc_framework/src/memory_tracker.c

```diff
@@ -9,112 +9,97 @@
  */
 
 #include "memory_tracker.h"
+#include "critnib/critnib.h"
+
+#include <umf/memory_pool.h>
 #include <umf/memory_provider.h>
 #include <umf/memory_provider_ops.h>
 
-#include <cassert>
-#include <map>
-#include <mutex>
-#include <shared_mutex>
+#include <assert.h>
+#include <errno.h>
 #include <stdlib.h>
 
-#ifdef _WIN32
-#include <windows.h>
-#endif
-
-// TODO: reimplement in C and optimize...
-struct umf_memory_tracker_t {
-    enum umf_result_t add(void *pool, const void *ptr, size_t size) {
-        std::unique_lock<std::shared_mutex> lock(mtx);
+#if !defined(_WIN32)
+critnib *TRACKER = NULL;
+void __attribute__((constructor)) createLibTracker() {
+    TRACKER = critnib_new();
+}
+void __attribute__((destructor)) deleteLibTracker() { critnib_delete(TRACKER); }
 
-        if (size == 0) {
-            return UMF_RESULT_SUCCESS;
-        }
+umf_memory_tracker_handle_t umfMemoryTrackerGet(void) {
+    return (umf_memory_tracker_handle_t)TRACKER;
+}
+#endif
 
-        auto ret =
-            map.try_emplace(reinterpret_cast<uintptr_t>(ptr), size, pool);
-        return ret.second ? UMF_RESULT_SUCCESS : UMF_RESULT_ERROR_UNKNOWN;
-    }
+struct tracker_value_t {
+    umf_memory_pool_handle_t pool;
+    size_t size;
+};
 
-    enum umf_result_t remove(const void *ptr, size_t size) {
-        std::unique_lock<std::shared_mutex> lock(mtx);
+static enum umf_result_t
+umfMemoryTrackerAdd(umf_memory_tracker_handle_t hTracker,
+                    umf_memory_pool_handle_t pool, const void *ptr,
+                    size_t size) {
+    assert(ptr);
 
-        map.erase(reinterpret_cast<uintptr_t>(ptr));
+    struct tracker_value_t *value =
+        (struct tracker_value_t *)malloc(sizeof(struct tracker_value_t));
+    value->pool = pool;
+    value->size = size;
 
-        // TODO: handle removing part of the range
-        (void)size;
+    int ret = critnib_insert((critnib *)hTracker, (uintptr_t)ptr, value, 0);
 
+    if (ret == 0) {
         return UMF_RESULT_SUCCESS;
     }
 
-    void *find(const void *ptr) {
-        std::shared_lock<std::shared_mutex> lock(mtx);
-
-        auto intptr = reinterpret_cast<uintptr_t>(ptr);
-        auto it = map.upper_bound(intptr);
-        if (it == map.begin()) {
-            return nullptr;
-        }
-
-        --it;
-
-        auto address = it->first;
-        auto size = it->second.first;
-        auto pool = it->second.second;
-
-        if (intptr >= address && intptr < address + size) {
-            return pool;
-        }
+    free(value);
 
-        return nullptr;
+    if (ret == ENOMEM) {
```

Comment:
There's also a possible `EEXIST` return value. Maybe we could return `UMF_RESULT_ERROR_INVALID_ARGUMENT` and log that the provided `ptr` argument already exists.

Reply:
Yeah, I left a comment that we should add some logging there. The problem is that right now we don't have a UMF logging framework and we shouldn't use the UR-one since those files will need to be moved to a different repo.